Examples/Algorithms/Generators/ActsExamples/Generators/EventGenerator.cpp

```diff
@@ -47,80 +88,483 @@ std::pair<std::size_t, std::size_t> EventGenerator::availableEvents() const {
 }
 
 ProcessCode EventGenerator::read(const AlgorithmContext& ctx) {
-  SimParticleContainer particles;
-  SimVertexContainer vertices;
+  ACTS_VERBOSE("EventGenerator::read");
+  std::vector<SimParticle> particlesUnordered;
+  std::vector<SimVertex> verticesUnordered;
 
   auto rng = m_cfg.randomNumbers->spawnGenerator(ctx);
 
+  std::vector<std::shared_ptr<HepMC3::GenEvent>> genEvents;
+
   std::size_t nPrimaryVertices = 0;
-  for (std::size_t iGenerate = 0; iGenerate < m_cfg.generators.size();
-       ++iGenerate) {
-    auto& generate = m_cfg.generators[iGenerate];
+  ACTS_VERBOSE("Using " << m_cfg.generators.size() << " generators");
+  {
+    Acts::AveragingScopedTimer genTimer("Generating primary vertices", logger(),
+                                        Acts::Logging::DEBUG);
 
-    // generate the primary vertices from this generator
-    for (std::size_t n = (*generate.multiplicity)(rng); 0 < n; --n) {
-      nPrimaryVertices += 1;
+    for (std::size_t iGenerate = 0; iGenerate < m_cfg.generators.size();
+         ++iGenerate) {
+      auto& generate = m_cfg.generators[iGenerate];
 
-      // generate primary vertex position
-      auto vertexPosition = (*generate.vertex)(rng);
-      // generate particles associated to this vertex
-      auto [newVertices, newParticles] = (*generate.particles)(rng);
-
-      ACTS_VERBOSE("Generate vertex at " << vertexPosition.transpose());
-
-      auto updateParticleInPlace = [&](SimParticle& particle) {
-        // only set the primary vertex, leave everything else as-is
-        // using the number of primary vertices as the index ensures
-        // that barcode=0 is not used, since it is used elsewhere
-        // to signify elements w/o an associated particle.
-        const auto pid = SimBarcode{particle.particleId()}.setVertexPrimary(
-            nPrimaryVertices);
-        // move particle to the vertex
-        const auto pos4 = (vertexPosition + particle.fourPosition()).eval();
-        ACTS_VERBOSE(" - particle at " << pos4.transpose());
-        // `withParticleId` returns a copy because it changes the identity
-        particle = particle.withParticleId(pid);
-        particle.initial().setPosition4(pos4);
-        particle.final().setPosition4(pos4);
-      };
-      for (auto& vertexParticle : newParticles) {
-        updateParticleInPlace(vertexParticle);
-      }
+      // generate the primary vertices from this generator
+      assert(generate.multiplicity != nullptr);
```

Comment:
guaranteed by construction?

Reply:
same below

Reply:
I actually already added the checks in the constructor. In that case, I'll remove these asserts.